### PR TITLE
User settings rewrite

### PR DESCRIPTION
**Pull request by evakill**
wanted to put this PR up even though the design may change a lil more

changed settings page to a tab view that includes the clubs membership settings, favorites, and user settings

eventually need to link admin pages to their console (change 'leave' button to 'manage')

### CODE REVIEW

File: components/Header/Links.js [hunk 1]
```diff
@@ -40,14 +40,8 @@ export default ({ userInfo, authenticated, show }) => (
...
       </StyledLink>
-      <StyledLink href="/favorites">Favorites</StyledLink>
-      {authenticated === false && (
-        <StyledLink
-          href={`${LOGIN_URL}?next=${window.location.href}`}
-          onClick={() => logEvent('login', 'click')}
-        >
-          Login
-        </StyledLink>
+      {(authenticated === false) && (
```
Comment (cacabo) [inline components/Header/Links.js]:
```suggestion
      {!authenticated && (
```

File: components/Settings/ClubTab.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import { DARK_GRAY } from '../../constants/colors'
+import { CLUB_ROUTE } from '../../constants/routes'
+import s from 'styled-components'
+import Toggle from './Toggle'
+import { Icon } from '../common'
+import ReactTooltip from 'react-tooltip'
+
+const Table = s.table`
```
Comment (cacabo) [inline components/Settings/ClubTab.js]: Should we put this along w/ the other shared components? Would be nice to use the same table styles everywhere

File: components/Settings/ClubTab.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  font-family: HelveticaNeue;
```
Comment (cacabo) [inline components/Settings/ClubTab.js]: Also would be good if fonts are a constant in another file (just like how colors are); then we could say:
```
font-family: ${BODY_FONT};
```
And can easily change fonts / not misspell them

File: components/Settings/ClubTab.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+              </td>
+              <td>
+                <Toggle club={item} active={item.public} toggle={(club) => togglePublic(club)}/>
+              </td>
+              <td>
+                { (item.role_display === 'Admin') ? <button className="button is-small">Manage</button> : <button className="button is-small" onClick={() => leaveClub(item)}>Leave</button> }
+              </td>
+            </tr>
+          )) : (
+            <tr>
```
Comment (cacabo) [inline components/Settings/ClubTab.js]: Might be better to put the error / not authed case as the first case in the ternary operator—it also might be better to not return a table at all and just return a non-ideal state type thing like what we do on the favorities page when the user has not favorited anything yet

File: components/Settings/ClubTab.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+              <td className='has-text-grey' colSpan='4'>
```
Comment (cacabo) [inline components/Settings/ClubTab.js]:
```suggestion
              <td className="has-text-grey" colSpan="4">
```

File: components/Settings/FavoritesTab.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import ClubTableRow from '../ClubTableRow'
+
+export default (props) => {
+  const { updateFavorites, favorites, clubs } = props
+
+  const findClub = clubs ? code => clubs.find(club => club.code === code) : {}
```
Comment (cacabo) [inline components/Settings/FavoritesTab.js]:
```suggestion
  const findClub = clubs ? code => clubs.find(club => club.code === code) : () => {}
```
Shouldn't it be a function in both cases?

File: components/Settings/FavoritesTab.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  return (
+    <div>
+      {favorites.map((favorite) => (
+        <ClubTableRow
+          club={findClub(favorite)}
```
Comment (cacabo) [inline components/Settings/FavoritesTab.js]: There should be a `key` prop which is unique for each favorite

File: components/Settings/FavoritesTab.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+          updateFavorites={updateFavorites}
+          openModal={null}
+          favorite={true}
+        />
+      ))}
+      {(!favorites.length) ? (
```
Comment (cacabo) [inline components/Settings/FavoritesTab.js]: I'd split these up into separate return statements.
```javascript
if (!favorites || !favorites.length) {
  return ...no favorites component
}

return ...normal component
```

File: components/Settings/ProfileTab.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import s from 'styled-components'
+
+const Card = s.div`
+  justify-content: space-between;
+  height: 100%;
+  font-family: HelveticaNeue;
```
Comment (cacabo) [inline components/Settings/ProfileTab.js]: same comment from above, would be nice to make fonts a constant in another file

File: components/Settings/ProfileTab.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  font-size: 18px;
+`
+
+export default (props) => {
+  const { name, username, email } = props.defaults
+
+  return (
+    <Card>
+      <div className='columns is-mobile'>
```
Comment (cacabo) [inline components/Settings/ProfileTab.js]:
```suggestion
      <div className="columns is-mobile">
```
use double quotes in jsx

File: components/Settings/ProfileTab.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        <div className='column is-narrow' style={{ fontWeight: 600 }}>
```
Comment (cacabo) [inline components/Settings/ProfileTab.js]:
```suggestion
        <div className="column is-narrow" style={{ fontWeight: 600 }}>
```

File: components/Settings/ProfileTab.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+          <div>Name:</div>
+          <div>Username:</div>
+          <div>Email:</div>
+        </div>
+        <div className='column is-narrow'>
```
Comment (cacabo) [inline components/Settings/ProfileTab.js]:
```suggestion
        <div className="column is-narrow">
```

File: components/Settings/ProfileTab.js [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+          <div>{name}</div>
+          <div>{username}</div>
+          <div>{email}</div>
+        </div>
+      </div>
+      <div style={{ fontSize: 12 }}>
+        If your information is incorrect, please send an email to <b>contact@pennclubs.com</b> detailing your issue.
```
Comment (cacabo) [inline components/Settings/ProfileTab.js]:
```suggestion
        If your information is incorrect, please send an email to <b><a href="mailto:contact@pennclubs.com">contact@pennclubs.com</a></b> detailing your issue.
```

File: components/Settings/Toggle.js [hunk 14]
```diff
@@ -0,0 +1,110 @@
+// modified from basics
```
Comment (cacabo) [inline components/Settings/Toggle.js]:
```suggestion
// Modified from basics
```

File: components/Settings/Toggle.js [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.25);
+  cursor: pointer;
+`
+
+/**
+ * @param {boolean} filter: filter in the redux
+ * @param {string} filterOffText: text rendered when the filter is off
+ * @param {string} filterOnText text rendered when filter is on
+ */
+
```
Comment (cacabo) [inline components/Settings/Toggle.js]:
```suggestion
```

File: components/TabView.js [hunk 16]
```diff
@@ -16,8 +16,8 @@ const TabView = ({ tabs }) => {
 
   return (
     <>
-      <div className="tabs">
-        <ul style={{ borderBottomWidth: '2px' }}>
+      <div className={'tabs ' + tabStyle}>
```
Comment (cacabo) [inline components/TabView.js]:
```suggestion
      <div className={`tabs  ${tabStyle}`}>
```

File: pages/settings.js [hunk 17]
```diff
@@ -1,23 +1,48 @@
...
+  font-family: HelveticaNeue;
+  font-size: 45px;
+  font-weight: bold;
+  color: ${DARK_GRAY};
+  margin-bottom: 1rem;
+
+`
+const Notification = s.span`
+  border-radius: 4px;
+  background-color: #c5e7f7;
```
Comment (cacabo) [inline pages/settings.js]: Would be good to not hardcode colors; can we get one from the colors file or add this color there?

File: pages/settings.js [hunk 18]
```diff
@@ -1,23 +1,48 @@
...
+  font-size: 16px;
+  font-weight: 500;
+  padding: 5px;
+  width: 500px;
```
Comment (cacabo) [inline pages/settings.js]: Probably don't want to hardcode width—might look odd on wide screens or on mobile. Probably need media queries here.

File: pages/settings.js [hunk 19]
```diff
@@ -1,23 +1,48 @@
...
+  float: right;
```
Comment (cacabo) [inline pages/settings.js]: Why use float here? Could be better to do
```
position: absolute;
right: 1rem;
```
Or something along those lines—lmk what you think works best!

File: pages/settings.js [hunk 20]
```diff
@@ -33,13 +58,19 @@ class SettingsForm extends React.Component {
...
   }
 
-  submit(data) {
-    doApiRequest('/settings/?format=json', {
-      method: 'PATCH',
-      body: data,
-    }).then(resp => {
+  togglePublic(club) {
+    doApiRequest(
+      `/clubs/${club.code}/members/${this.props.userInfo.username}/?format=json`,
```
Comment (cacabo) [inline pages/settings.js]: can you put something like
```javascript
const { userInfo: { username }} = this.props
```
before this string interpolation to cut down on how long it is and make it a little easier to read?

File: pages/settings.js [hunk 21]
```diff
@@ -69,174 +100,92 @@ class SettingsForm extends React.Component {
...
-                </table>
-              </div>
-            ),
-          },
-        ],
-      },
-    ]
+  leaveClub(club) {
+    if (confirm(`Are you sure you want to leave ${club.name}? You cannot add yourself back into the club.`)) {
+      doApiRequest(`/clubs/${club.code}/members/${this.props.userInfo.username}`, {
```
Comment (cacabo) [inline pages/settings.js]: Same thing here—would be good to parse out the `username` before putting it in the string to cut down on length + improve readability (also might want to do some error checking to make sure that the username is present)

File: pages/settings.js [hunk 22]
```diff
@@ -69,174 +100,92 @@ class SettingsForm extends React.Component {
...
-          onSubmit={this.submit}
-        />
-        <a
-          href={`${API_BASE_URL}/accounts/logout/?next=${window.location.href}`}
-          className="button is-pulled-right is-danger is-medium"
-        >
-          Logout
-        </a>
+          </Notification>
+        ) : ''}
```
Comment (cacabo) [inline pages/settings.js]:
```suggestion
        ) : null}
```

File: renderPage.js [hunk 23]
```diff
@@ -108,6 +93,28 @@ function renderPage(Page) {
...
+            this.setState({
+              authenticated: true,
+              favorites: data.favorite_set.map(a => a.club),
+              userInfo: data,
+            })
+          })
+        } else {
+          this.setState({
+            authenticated: false,
+            favorites: JSON.parse(localStorage.getItem('favorites')) || [],
```
Comment (cacabo) [inline renderPage.js]: Would be good to abstract all operations on local store to their own file and we just call a function like `store.getFavorites()`

If you'd rather not do that in this PR, can you make a Trello card so we make sure we do it eventually?

File: renderPage.js [hunk 24]
```diff
@@ -156,15 +154,13 @@ export function renderListPage(Page) {
...
       }
 
-      const favoriteClubs = this.mapToClubs(favorites)
-
       return (
         <Page
           clubs={clubs}
           tags={tags}
           favorites={favorites}
           updateFavorites={this.props.updateFavorites}
```
Comment (cacabo) [inline renderPage.js]: Would be nice to destructure props before render:
```js
const { updateFavorites, updateUserInfo } = this.props
...
return (...)
```

File: renderPage.js (hunk 23 above)
Comment (evakill) [reply renderPage.js]: to trello

File: components/Settings/ClubTab.js (hunk 2 above)
Comment (evakill) [reply components/Settings/ClubTab.js]: i'm going to put this off, we don't have tables elsewhere and passing in the icon and tooltip will be annoyingggggg

**Comment by evakill** [comment]
look whack on mobile but that's a problem for another day

**Comment by evakill** [comment]
Screen Shot 2019-11-18 at 11 09 06 PM
Screen Shot 2019-11-18 at 11 09 10 PM
Screen Shot 2019-11-18 at 11 09 14 PM

**Comment by cacabo** [comment]
cool just made a card for fixing mobile jawn. lgtm. good work boss.

**Pull request merged**